src/Utilities/Compiler/SmallDictionary.cs

```diff
@@ -0,0 +1,861 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Analyzer.Utilities
+{
+    /// <summary>
+    /// Dictionary designed to hold small number of items.
+    /// Compared to the regular Dictionary, average overhead per-item is roughly the same, but 
+    /// unlike regular dictionary, this one is based on an AVL tree and as such does not require 
+    /// rehashing when items are added.
+    /// It does require rebalancing, but that is allocation-free.
+    ///
+    /// Major caveats:
+    ///  1) There is no Remove method. (can be added, but we do not seem to use Remove that much)
+    ///  2) foreach [keys|values|pairs] may allocate a small array.
+    ///  3) Performance is no longer O(1). At a certain count it becomes slower than regular Dictionary.
+    ///     In comparison to regular Dictionary on my machine:
+    ///        On trivial number of elements (5 or so) it is more than 2x faster.
+    ///        The break even count is about 120 elements for read and 55 for write operations (with unknown initial size).
+    ///        At UShort.MaxValue elements, this dictionary is 6x slower to read and 4x slower to write
+    ///
+    /// Generally, this dictionary is a win if number of elements is small, not known beforehand or both.
+    ///
+    /// If the size of the dictionary is known at creation and it is likely to contain more than 10 elements, 
+    /// then regular Dictionary is a better choice.
+    /// </summary>
+#pragma warning disable CA1051, CA1716 // Do not declare visible instance fields
+    internal sealed class SmallDictionary<K, V> : IEnumerable<KeyValuePair<K, V>>
```

Comment:
Out of curiosity, where did this come from?

Reply:
https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Collections/SmallDictionary.cs - this is the highest allocating data structure in the analyzer for holding decoded platform attributes data, so @buyaa-n was looking for an efficient dictionary implementation when it is guaranteed to hold limited set of entries. I pointed her to one used in Roslyn for such cases.

Reply:
@danmosemsft now that's a specialized set of collections we should productize :-)

Reply:
Yeah, I have also seen small lists as well eg 
https://github.com/dotnet/msbuild/blob/f6739274e179fd1ecb7317b8fdc6a3e2c09b4ee0/src/Build/Collections/MultiDictionary.cs#L165

But I need your help devising that working model for shipping useful collections that aren't critical to the core libraries ... 😃